module.nix

```diff
@@ -0,0 +1,295 @@
+      };
+
+      host = mkOption {
+        type = types.str;
+        default = "127.0.0.1";
+        description = "Hostname of the database server";
+      };
+
+      port = mkOption {
+        type = types.port;
+        default = 5432;
+        description = "Postgresql database port. Must be correct even if `services.teslamate.postgres.enable` is false";
+      };
+    };
+
+    grafana = {
+      enable = mkOption {
+        type = types.bool;
+        default = false;
+        description = "Whether to create and provision grafana with the TeslaMate dashboards";
+      };
+      
+      listenAddress = mkOption {
+        type = types.str;
+        default = "0.0.0.0";
+        description = "IP address for grafana to listen to.";
+      };
+
+      port = mkOption {
+        type = types.port;
+        default = 3000;
+        description = "Port for grafana web service";
+      };
+
+      urlPath = mkOption {
+        type = types.str;
+        default = "/";
+        description = "Path that grafana is mounted on. Useful if using a reverse proxy to vend teslamate and grafana on the same port";
+      };
+    };
+
+    mqtt = {
+      enable = mkEnableOption "TeslaMate MQTT integration";
+
+      host = mkOption {
+        type = types.str;
+        default = "127.0.0.1";
+        description = "MQTT host";
+      };
+
+      port = mkOption {
+        type = with types; nullOr port;
+        default = null;
+        example = 1883;
+        description = "MQTT port.";
+      };
+    };
+  };
+
+  config = mkIf cfg.enable 
+    (mkMerge [
+      {
+        users.users.teslamate = {
+          isSystemUser = true;
+          group = "teslamate";
+          home = "/var/lib/teslamate";
+          createHome = true;
+        };
+        users.groups.teslamate = {};
+
+        systemd.services.teslamate = {
+          description = "TeslaMate";
+          after = [ "network.target" "postgresql.service" ];
+          wantedBy = mkIf cfg.autoStart [ "multi-user.target" ];
+          serviceConfig = {
+            User = "teslamate";
+            Restart = "on-failure";
+            RestartSec = 5;
+
+            WorkingDirectory = "/var/lib/teslamate";
+
+            ExecStartPre = ''${getExe teslamate} eval "TeslaMate.Release.migrate"'';
+            ExecStart = "${getExe teslamate} start";
+            ExecStop = "${getExe teslamate} stop";
+
+            EnvironmentFile = cfg.secretsFile;
+          };
+          environment = mkMerge [
+            {
+              PORT = toString cfg.port;
+              DATABASE_USER = cfg.postgres.user;
+              DATABASE_NAME = cfg.postgres.database;
+              DATABASE_HOST = cfg.postgres.host;
+              DATABASE_PORT = toString cfg.postgres.port;
+              VIRTUAL_HOST = cfg.virtualHost;
+              URL_PATH = cfg.urlPath;
+              HTTP_BINDING_ADDRESS = mkIf (cfg.listenAddress != null) cfg.listenAddress;
+              DISABLE_MQTT = mkIf (!cfg.mqtt.enable) "true";
```

Comment:
I am a bit uneasy that `postgres.enable` enables/disables the postgres server, but `mqtt.enable` enables functionality on the client. This could be a bit confusing.

Maybe `postgresl.enable_server` might be better here? As we still configure the postgres server in the client even if it is disabled.

Reply:
Yeah that's a fair critique. I felt very awkward about that as well. Perhaps `postgresql.create_server`? Or `postgresql.setup_server`? Or would it be better to take the opposite approach and rename `mqtt.enable` to `mqtt.disable` to align it with the actual env var that will get set? Thoughts?

Reply:
Thinking about this some more, I think I would go for `enable_server`.

Leaving `mqtt.enable` is more in line with my expectations from other projects. And same with `grafana.enable` also.

Not extremely fussed though.

Reply:
Ok that sounds reasonable. I will go with `enable_server`